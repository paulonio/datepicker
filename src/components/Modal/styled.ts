import styled from 'styled-components';

export const Wrapper = styled.div`
  margin-top: 8px;
  width: 250px;
  padding: 10px;
  border: 1px solid #333333;
  border-radius: 4px;
`;

export const Title = styled.h3`
  color: #333333;
  font-size: 14px;
  font-weight: 700;
  line-height: 19px;
`;

export const InputWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const Input = styled.input`
  max-width: 140px;
`;

export const Button = styled.button`
  max-width: 100px;
  &:hover {
    cursor: pointer;
  }
`;
