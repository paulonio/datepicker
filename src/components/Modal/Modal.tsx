import React, { ChangeEvent, FC, useEffect, useState } from 'react';
import { Button, Input, Title, Wrapper } from './styled';
import TasksList from '../TasksList/TasksList';

interface ModalProps {
  date: Date;
  handleCloseModal: (value: boolean) => void;
}

export interface Task {
  id: string;
  title: string;
}

const getTasks = (date: Date) => {
  const tasks = JSON.parse(localStorage.getItem(`${date}`) as string);
  return tasks;
};

const Modal: FC<ModalProps> = ({ date, handleCloseModal }) => {
  const [value, setValue] = useState<string>('');
  const [tasks, setTasks] = useState<Task[]>([]);

  useEffect(() => {
    const receivedTasks = getTasks(date);
    if (receivedTasks) {
      setTasks(receivedTasks);
    } else {
      setTasks([]);
    }
  }, [date]);

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const inputValue = e.target.value;
    setValue(inputValue);
  };

  const handleAddTask = () => {
    if (!localStorage.getItem(`${date}`)) {
      localStorage.setItem(`${date}`, JSON.stringify([{ id: '0', title: value }]));
    } else {
      const prevValues: Array<Task> = JSON.parse(localStorage.getItem(`${date}`) as string);
      const newTask: Task = { id: `${prevValues.length}`, title: value };
      const data = [...prevValues, newTask];
      localStorage.setItem(`${date}`, JSON.stringify(data));
    }
    setValue('');
    handleCloseModal(false);
  };

  return (
    <Wrapper>
      <Title>Add task</Title>
      <Input value={value} onChange={handleChange} placeholder="Type task" />
      <Button onClick={handleAddTask}>Confirm</Button>
      <TasksList tasks={tasks} date={date} />
    </Wrapper>
  );
};

export default Modal;
