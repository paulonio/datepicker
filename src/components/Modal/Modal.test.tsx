import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import Modal from '@components/Modal/Modal';

describe('Modal', () => {
  const dateMock = new Date(2023, 4, 11);
  jest.spyOn(Storage.prototype, 'setItem');
  Storage.prototype.setItem = jest.fn();

  it('should render modal component', () => {
    const { container } = render(<Modal date={dateMock} handleCloseModal={() => {}} />);

    expect(container).toMatchSnapshot();
  });

  it('should change input', () => {
    render(<Modal date={dateMock} handleCloseModal={() => {}} />);

    const input: HTMLInputElement = screen.getByTestId('input');
    expect(input).toHaveTextContent('');
    fireEvent.change(input, { target: { value: 'Learn react' } });
    expect(input.value).toBe('Learn react');
  });

  it('should set task to localStorage', () => {
    render(<Modal date={dateMock} handleCloseModal={() => {}} />);

    const input: HTMLInputElement = screen.getByTestId('input');
    const submitButton = screen.getByTestId('submit');
    fireEvent.change(input, { target: { value: 'Learn react' } });
    fireEvent.click(submitButton);
    expect(input.value).toBe('');
    expect(localStorage.setItem).toHaveBeenCalled();
  });
});
