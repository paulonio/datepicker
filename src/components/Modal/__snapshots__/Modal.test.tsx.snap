// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render modal component 1`] = `
<div>
  <div
    class="sc-fFGjHI gllHxj"
    data-testid="modal"
  >
    <h3
      class="sc-dicizt bieuVL"
    >
      Add task
    </h3>
    <div
      class="sc-iVCKna kYeVvR"
    >
      <input
        class="sc-gxYJeL hyoWFD"
        data-testid="input"
        placeholder="Type task"
        value=""
      />
      <button
        class="sc-kPTPQs gXqdmt"
        data-testid="submit"
      >
        Confirm
      </button>
    </div>
    <div
      class="sc-bgqQcB eaEnve"
    >
      <h3
        class="sc-gTRrQi gteUIR"
        data-testid="title"
      >
        Tasks - 
        11/05/2023
      </h3>
    </div>
  </div>
</div>
`;
